src/geopackage/read.cpp

```diff
@@ -0,0 +1,108 @@
+#include "GeoPackage.hpp"
+
+#include <numeric>
+
+std::shared_ptr<geojson::FeatureCollection> geopackage::read(
+    const std::string& gpkg_path,
+    const std::string& layer = "",
+    const std::vector<std::string>& ids = {}
+)
+{
+    sqlite db(gpkg_path);
+
+    // Check if layer exists
+    if (!db.has_table(layer)) {
+        // Since the layer doesn't exist, we need to output some additional
+        // debug information with the error. In this case, we add ALL the tables
+        // available in the GPKG, so that if the user sees this error, then it
+        // might've been either a typo or a bad data input, and they can correct.
+        std::string errmsg = "[" + std::string(sqlite3_errmsg(db.connection())) + "] " +
+                             "table " + layer + " does not exist.\n\tTables: ";
+
+        auto errquery = db.query("SELECT name FROM sqlite_master WHERE type='table'").next();
+        while(!errquery.done()) {
+            errmsg += errquery.get<std::string>(0);
+            errmsg += ", ";
+            errquery.next();
+        }
+
+        throw std::runtime_error(errmsg);
+    }
+
+    // Layer exists, getting statement for it
+    //
+    // this creates a string in the form:
+    //     WHERE id IN (?, ?, ?, ...)
+    // so that it can be bound by SQLite.
+    // This is safer than trying to concatenate
+    // the IDs together.
+    std::string joined_ids = "";
+    if (!ids.empty()) {
+        joined_ids = " WHERE id IN (?";
+        for (size_t i = 1; i < ids.size(); i++) {
+            joined_ids += ", ?";
+        }
+        joined_ids += ")";
+    }
+
+    // Get number of features
+    sqlite_iter query_get_layer_count = db.query("SELECT COUNT(*) FROM " + layer + joined_ids, ids);
```

Comment:
Is `layer` something we potentially need to worry about as untrusted input (e.g. if there's some web front-end for launching a simulation)? I see that it's passed as a query parameter in the implementation of `has_table`, but concatenated into the query string here. I would *not* rely on the affirmative return from `has_table` to tell us that it's safe to use like this.

Reply:
It's also parameterized down below, on line 67

Reply:
And then concatenated again on 72

Reply:
Table names can't be parameterized in the `FROM` clause, as far as I know. The one below parameterizes `layer` as part of the `WHERE` clause, which is the same in `has_table`.

Reply:
In that case, if `layer` might be untrusted input, we need to very strictly limit its contents to a sequence of characters that represent a single valid (potential) table name and nothing more.

Reply:
I won't hold things up over this, since I have the least context to make a good decision.

If we do want to sanitize `layer`, I'd propose a two-layer white-list check before the `has_table` call up above
* composed only of alpha-numeric characters, `-`, and `_` (i.e. matches regular expression `[-A-Za-z0-9_]+`), and
* does not begin with `sqlite_` (because SQLite uses that for special internal tables)

I omitted `.` from my suggestion of acceptable characters because that's what SQLite uses for a "schema name" separator. We could also maybe work around that by always prefixing `layer` as a table name with a scheme name that we specify, so that subsequent `.` don't have any additional semantic.

Reply:
I've implemented the two-layer white-list check in 8cbd253. Before the `has_table` call, a check for `sqlite_` and invalid characters is done in a call to the function `check_table_name`.